### PR TITLE
Henryh/add dino v3

### PR DESCRIPTION
**Pull request by Hgherzog**
PR Summary: DINOv3 Integration & Evaluation Pipeline Enhancements
🚀 Major Features
DINOv3 Model Integration: Complete implementation of DINOv3 model wrapper with support for MaskedHeliosSample objects, multiple modalities, and normalization
Enhanced Normalization System: New normalization methods with min/max statistics tracking, bounds calculation, and flexible post-processing options
Galileo: Added pre-train normalization
Panopticon: Updated defaults 

🔧 Core Changes
Evaluation Pipeline:
Refactored all_evals.py with improved type hints and configuration management
Added full evaluation sweep functionality with parameter sweeps for learning rate, normalization modes, and pooling types
Enhanced linear probe training to handle segmentation tasks with varying image sizes via patch-based logit normalization
Data Processing:
Updated GeobenchDataset to track min/max normalization statistics alongside mean/std
Added Normalizer class for band normalization with configurable normalization values
Improved CropHarvestDataset to not require latlon to be output for effective parsing
🧪 Testing & Infrastructure
Comprehensive test suite for evaluation sweep command builder
New unit tests for normalization functions covering edge cases
Reorganized test structure (moved eval tests to proper directories)
Added fixtures for reusable test data
📝 Configuration & Documentation
Updated README with budget information and evaluation instructions
Added normalization configurations for Galileo model features
Updated dataset configurations to handle unrecognized datasets efficiently
🎯 Review Focus Areas
DINOv3 Integration (helios/evals/models/dinov3/) - New model wrapper implementation
Normalization Changes (helios/evals/) - Enhanced normalization system with backwards compatibility
Linear Probe updates (helios/evals/llinear_probe.py) - Computing logits per patch from hw and number of spatial embeddings rather than patch size to work with models that need resizing.
Evaluation Pipeline (helios/internal/full_eval_sweep.py) - New sweep functionality
Test Coverage (tests/) - Comprehensive test additions for new features
⚠️ Breaking Changes
Updated function signatures in evaluation modules (added norm_method parameters)
Refactored linear probe evaluation logic for segmentation tasks
This PR significantly expands the model ecosystem (adding DINOv3), improves data processing flexibility, and strengthens the evaluation infrastructure with comprehensive testing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -96,3 +96,6 @@ We have 2 versions of each dataset 1 with 256 x 256 tiles and 1 with 4x as many
 - **WorldCover Sampling Dataset**: ~1.6M samples with Landsat, OpenStreetMap raster, Sentinel-1, Sentinel-2 L2A, SRTM, and WorldCover modalities. WorldCover class based sampling and some additional random sampling over the rest of the world.
   - 256 Path: `/weka/dfive-default/helios/dataset/worldcover_sampling/h5py_data_w_missing_timesteps_zstd_3/landsat_openstreetmap_raster_sentinel1_sentinel2_l2a_srtm_worldcover/1592645/`
   - 128 Path: `/weka/dfive-default/helios/dataset/worldcover_sampling/h5py_data_w_missing_timesteps_zstd_3_128_x_4/landsat_openstreetmap_raster_sentinel1_sentinel2_l2a_srtm_worldcover/6370580/`
+
+
+## Running Eval Suite
```
Comment (Hgherzog) [inline README.md:102]: Fill in this information so it is obvious how to run evals

File: helios/evals/datasets/__init__.py [hunk 2]
```diff
@@ -28,12 +31,40 @@ class EvalDatasetPartition(StrEnum):
     TRAIN_050X = "0.50x_train"
 
 
+# Supported Modalities per Eval Dataset
+# TODO: Add more datasets
```
Comment (Hgherzog) [inline helios/evals/datasets/__init__.py]: Move this to just be stored in the dataset config

File: helios/evals/datasets/normalize.py [hunk 3]
```diff
@@ -44,18 +47,31 @@ class NormMethod(str, Enum):
...
+    NORM_YES_CLIP_2_STD = "norm_yes_clip_2_std"
+    NORM_YES_CLIP_3_STD_INT = "norm_yes_clip_3_std_int"
+    NORM_YES_CLIP_2_STD_INT = "norm_yes_clip_2_std_int"
     NORM_YES_CLIP_INT = "norm_yes_clip_int"
+    NORM_YES_CLIP_MIN_MAX_INT = "norm_yes_clip_min_max_int"
     STANDARDIZE = "standardize"
     NO_NORM = "no_norm"
 
-
+# TODO: Add a unit test for this
```
Comment (Hgherzog) [inline helios/evals/datasets/normalize.py]: Add unit tests for this function

File: helios/evals/datasets/normalize.py [hunk 4]
```diff
@@ -44,18 +47,31 @@ class NormMethod(str, Enum):
...
     """Normalize an image with given mean and std arrays, and a normalization method."""
+    if isinstance(method, str):
+        method = NormMethod(method)
+    if mins is None:
+        logger.info("No mins provided, falling back to 2 stds int")
+        # Hack for now for dino for non geobench datasets
+        method = NormMethod.NORM_YES_CLIP_2_STD_INT
     original_dtype = image.dtype
     if method == NormMethod.NO_NORM:
         print("No normalization")
```
Comment (Hgherzog) [inline helios/evals/datasets/normalize.py]: log instead of print

File: helios/evals/datasets/normalize.py [hunk 5]
```diff
@@ -44,18 +47,31 @@ class NormMethod(str, Enum):
...
+    means: np.ndarray,
+    stds: np.ndarray,
+    mins: np.ndarray | None = None,
+    maxs: np.ndarray | None = None,
     method: str = NormMethod.NORM_NO_CLIP,
 ) -> np.ndarray:
     """Normalize an image with given mean and std arrays, and a normalization method."""
+    if isinstance(method, str):
+        method = NormMethod(method)
+    if mins is None:
```
Comment (Hgherzog) [inline helios/evals/datasets/normalize.py]: Oh no this is messing up the results for panopticon and Galileo on non-geobench datasets which is weird. We need to find a better way to fallback if mins are not provided, or just go calculate all the mins and maxs to try and do this fairly for all the datasets.

File: helios/evals/linear_probe.py [hunk 6]
```diff
@@ -236,6 +236,7 @@ def train_probe(
                 logits = outputs["logits"]
                 # logger.info(f"logits: {logits.shape}")
                 if task_type == TaskType.SEGMENTATION:
+                    # This is effectively nearest neighbor interpolation
```
Comment (Hgherzog) [inline helios/evals/linear_probe.py:251]: Given the patch size and the output shape I want to be clear about how we calculate the probe output size and do this re-shaping

File: helios/evals/models/__init__.py [hunk 7]
```diff
@@ -1,14 +1,19 @@
 """Models for evals."""
 
 from helios.evals.models.dinov2.dinov2 import DINOv2, DINOv2Config
+from helios.evals.models.dinov3.dinov3 import DINOv3, DINOv3Config
 from helios.evals.models.galileo import GalileoConfig, GalileoWrapper
 from helios.evals.models.panopticon.panopticon import Panopticon, PanopticonConfig
 
+
+#TODO: assert that they all store a patch_size variable and supported modalities
```
Comment (Hgherzog) [inline helios/evals/models/__init__.py]: Adress TODO

File: helios/evals/models/dinov2/dinov2.py [hunk 8]
```diff
@@ -115,6 +115,7 @@ def _process_modality_data(
             data_i = F.interpolate(
                 data_i,
                 size=(new_height, new_height),
+                # maybe i should use the reccomended transform here
```
Comment (Hgherzog) [inline helios/evals/models/dinov2/dinov2.py:118]: Update this code to use same normalization procedure as dinov3

File: helios/evals/models/dinov3/constants.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    BASE_WEB = "dinov3_vitb16"
+    LARGE_WEB = "dinov3_vitl16"
+    HUGE_PLUS_WEB = "dinov3_vith16plus"
+    FUll_7B_WEB = "dinov3_vit7b16"
+    LARGE_SATELLITE = "dinov3_vitl16_sat"  # just the name I copied too
+    FULL_7B_SATELLITE = "dinov3_vit7b16_sat"
+    # Not yet clear how to download the satelite models
+
+
+# if non Ai2, user set the DINOV3_REPO_DIR and DINOV3_CHECKPOINT_DIR environment variables
```
Comment (Hgherzog) [inline helios/evals/models/dinov3/constants.py:21]: Make sure this information is included in an easy to find instructions document for running evals

File: helios/train/callbacks/evaluator_callback.py [hunk 10]
```diff
@@ -58,6 +63,7 @@ def __init__(
         task: DownstreamTaskConfig,
         trainer: Trainer,
         device: torch.device | None = None,
+        use_patch_size_from_model: bool = False,
```
Comment (Hgherzog) [inline helios/train/callbacks/evaluator_callback.py]: Not sure if it is clear what this is doing needs a better name or arg comment or a way to derive it given the situation

File: helios/train/callbacks/evaluator_callback.py [hunk 11]
```diff
@@ -183,14 +201,18 @@ def val(self) -> float:
...
 
-        val_result = self.eval_function(  # type: ignore
-            config=self.config,
-            train_embeddings=train_embeddings,
-            train_labels=train_labels,
-            test_embeddings=test_embeddings,
-            test_labels=test_labels,
-            device=self.device,
-        )
+        # HACK Until I know what to do about the patch size for the segmentation head output
```
Comment (Hgherzog) [inline helios/train/callbacks/evaluator_callback.py]: Re: Patch size issue

File: helios/train/callbacks/evaluator_callback.py [hunk 12]
```diff
@@ -220,9 +243,44 @@ def pre_train(self) -> None:
                 for callback in self.trainer._iter_callbacks()
                 if isinstance(callback, HeliosWandBCallback)
             )
-            if wandb_callback.enabled:
-                for evaluator in self.evaluators:
-                    val_result, eval_time = self._perform_eval(evaluator)
+            if hasattr(self.trainer.train_module.model, "supported_modalities"):
```
Comment (Hgherzog) [inline helios/train/callbacks/evaluator_callback.py]: Make a separate function for determining if the modalities are supported and the eval should be skipped

File: scripts/dino_v3/dino_v3_launch.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+from helios.evals.models.dinov3.dinov3 import DinoV3Models
+from helios.internal.experiment import CommonComponents, HeliosVisualizeConfig
+from helios.nn.latent_mim import LatentMIMConfig
+from helios.train.loss import LossConfig
+from helios.train.masking import MaskingConfig
+from helios.train.train_module.latent_mim import LatentMIMTrainModuleConfig
+
+logger = logging.getLogger(__name__)
+
+MAX_PATCH_SIZE = 8
```
Comment (Hgherzog) [inline scripts/dino_v3/dino_v3_launch.py]: Remove unused parts of this

File: scripts/run_all_evals/all_evals.py [hunk 14]
```diff
@@ -45,20 +44,21 @@ def load_user_module(path):
...
-    "m_forestnet": DownstreamTaskConfig(
-        dataset="m-forestnet",
+    "m_eurosat": DownstreamTaskConfig(
+        dataset="m-eurosat",
         embedding_batch_size=128,
-        num_workers=4,
+        num_workers=0,
         pooling_type=PoolingType.MEAN,
-        norm_stats_from_pretrained=False,
+        norm_stats_from_pretrained=False,  # True, #False,
```
Comment (Hgherzog) [inline scripts/run_all_evals/all_evals.py]: Revert to true

File: scripts/run_all_evals/full_eval_sweep.py [hunk 15]
```diff
@@ -15,8 +16,8 @@
 from helios.nn.flexihelios import PoolingType
 
 LP_LRs = [1e-4, 5e-4, 1e-3, 5e-3, 1e-2, 5e-2, 1e-1, 5e-1]
-Normalization_MODES = ["dataset", "helios"]
-pooling_types = [PoolingType.MAX, PoolingType.MEAN]
+Normalization_MODES = ["dataset", "pre_trained"]
+pooling_types = [PoolingType.MEAN, PoolingType.MAX]
 
 logger = getLogger(__name__)
```
Comment (Hgherzog) [inline scripts/run_all_evals/full_eval_sweep.py]: I need to test and clean up this cli especially for running the other models so it is clear what to do

File: helios/evals/datasets/normalize.py (hunk 5 above)
Comment (Hgherzog) [reply helios/evals/datasets/normalize.py]: Best Solution is to calculate the min and max values as well. Panopticon and galileo need to be-re ran on pastis, mados, sickle0landsat and breizcrops

File: helios/evals/datasets/normalize.py [hunk 16]
```diff
@@ -44,18 +47,32 @@ class NormMethod(str, Enum):
...
+    means: np.ndarray,
+    stds: np.ndarray,
+    mins: np.ndarray | None = None,
+    maxs: np.ndarray | None = None,
     method: str = NormMethod.NORM_NO_CLIP,
 ) -> np.ndarray:
     """Normalize an image with given mean and std arrays, and a normalization method."""
+    if isinstance(method, str):
+        method = NormMethod(method)
+    if mins is None and method == NormMethod.NORM_YES_CLIP_MIN_MAX_INT:
```
Comment (Hgherzog) [inline helios/evals/datasets/normalize.py]: Leaving this issue as a TODO: https://github.com/allenai/helios/issues/334

File: scripts/dino_v2_evals/dino_v2.py
Comment (Hgherzog) [inline scripts/dino_v2_evals/dino_v2.py:1]: Leaving this file in but moved a launch script to maintain to the main codebase

File: scripts/dino_v2_evals/galileo.py
Comment (Hgherzog) [inline scripts/dino_v2_evals/galileo.py:1]: Leaving this file in but moved a launch script to maintain to the main codebase

File: scripts/dino_v2_evals/panopticon.py
Comment (Hgherzog) [inline scripts/dino_v2_evals/panopticon.py:1]: Leaving this file in but moved a launch script to maintain to the main codebase

File: scripts/dino_v3/dino_v3_launch.py
Comment (Hgherzog) [inline scripts/dino_v3/dino_v3_launch.py:1]: Leaving this file in but moved a launch script to maintain to the main codebase, open to deleting

File: scripts/dino_v2_evals/panopticon.py
Comment (Hgherzog) [reply scripts/dino_v2_evals/panopticon.py:1]: open to deleting

File: scripts/dino_v2_evals/galileo.py
Comment (Hgherzog) [reply scripts/dino_v2_evals/galileo.py:1]: open to deleting

File: README.md [hunk 17]
```diff
@@ -96,3 +97,80 @@ We have 2 versions of each dataset 1 with 256 x 256 tiles and 1 with 4x as many
...
+```bash
+python3 helios/internal/full_eval_sweep.py \
+  --cluster=ai2/saturn-cirrascale \
+  --galileo \
+  --model_name=galileo_vit_base
+  --model.patch_size=4
+```
+
+**Key Notes:**
+- The script automatically determines appropriate normalization strategies for each model type
```
Comment (gabrieltseng) [inline README.md]: Could you explicitly say what these are? (dataset and pretraining)

File: helios/evals/datasets/configs.py [hunk 18]
```diff
@@ -97,51 +107,61 @@ class EvalDatasetConfig(NamedTuple):
...
+    ),
+    "cropharvest": EvalDatasetConfig(
+        task_type=TaskType.CLASSIFICATION,
+        imputes=[
+            ("02 - Blue", "01 - Coastal aerosol"),
+            ("11 - SWIR", "10 - SWIR - Cirrus"),
+        ],
+        num_classes=2,
+        is_multilabel=False,
+        supported_modalities=[Modality.SENTINEL2_L2A.name, Modality.SENTINEL1.name],
```
Comment (gabrieltseng) [inline helios/evals/datasets/configs.py]: also `Modality.SRTM.name`

File: helios/evals/datasets/configs.py [hunk 19]
```diff
@@ -97,51 +107,61 @@ class EvalDatasetConfig(NamedTuple):
...
-                ("11 - SWIR", "10 - SWIR - Cirrus"),
-            ],
-            num_classes=2,
-            is_multilabel=False,
-        )
-    else:
+    key = "cropharvest" if dataset.startswith("cropharvest") else dataset
+    try:
+        return DATASET_TO_CONFIG[key]
+    except KeyError:
```
Comment (gabrieltseng) [inline helios/evals/datasets/configs.py:170]: this is nicer thanks

File: helios/evals/datasets/geobench_dataset.py [hunk 20]
```diff
@@ -146,7 +146,7 @@ def __init__(
             config.imputes,
             all_bands=GEOBENCH_L8_BAND_NAMES if self.is_landsat else EVAL_S2_BAND_NAMES,
         )
-        self.mean, self.std = self._get_norm_stats(
+        self.mean, self.std, self.min, self.max = self._get_norm_stats(
```
Comment (gabrieltseng) [inline helios/evals/datasets/geobench_dataset.py:149]: Do we need to update the other eval datasets to also compute min / max ?

File: helios/evals/datasets/normalize.py [hunk 21]
```diff
@@ -44,42 +47,102 @@ class NormMethod(str, Enum):
 
     NORM_NO_CLIP = "norm_no_clip"
     NORM_YES_CLIP = "norm_yes_clip"
+    NORM_YES_CLIP_3_STD = "norm_yes_clip_3_std"
+    NORM_YES_CLIP_2_STD = "norm_yes_clip_2_std"
+    NORM_YES_CLIP_3_STD_INT = "norm_yes_clip_3_std_int"
+    NORM_YES_CLIP_2_STD_INT = "norm_yes_clip_2_std_int"
```
Comment (gabrieltseng) [inline helios/evals/datasets/normalize.py:56]: these "stretched" stds were only used in the Galileo sweep - I don't think (?) any models use them, so if we aren't sweeping norm methods (beyond pretraining + dataset) they may not be necessary.

File: helios/evals/datasets/geobench_dataset.py (hunk 20 above)
Comment (gabrieltseng) [reply helios/evals/datasets/geobench_dataset.py:149]: I see: https://github.com/allenai/helios/issues/334

File: tests/unit/eval/datasets/test_normalize.py [hunk 22]
```diff
@@ -0,0 +1,394 @@
...
+        assert result == band_info
+
+    def test_no_missing_bands_returns_original(self) -> None:
+        """Test when all bands are present, no imputation needed."""
+        band_info = {
+            "01 - Coastal aerosol": {"mean": 0.1, "std": 0.02},
+            "02 - Blue": {"mean": 0.2, "std": 0.03},
+        }
+        all_bands = ["01 - Coastal aerosol", "02 - Blue"]
+        imputes = [("01 - Coastal aerosol", "02 - Blue")]
```
Comment (gabrieltseng) [inline tests/unit/eval/datasets/test_normalize.py]: Do we want to raise an error or a warning if the user expect something to be imputed but its not / doesn't need to be? it would imply someone is making a mistake.

File: tests/unit/eval/datasets/test_normalize.py [hunk 23]
```diff
@@ -0,0 +1,394 @@
...
+    def test_impute_order_matters(self) -> None:
+        """Test that first matching impute is used when multiple imputes target same band."""
+        band_info = {
+            "01 - Coastal aerosol": {"mean": 0.1, "std": 0.02},
+            "03 - Green": {"mean": 0.3, "std": 0.04},
+        }
+        all_bands = ["01 - Coastal aerosol", "02 - Blue", "03 - Green"]
+        imputes = [
+            ("01 - Coastal aerosol", "02 - Blue"),
+            ("03 - Green", "02 - Blue"),  # This should not be used
```
Comment (gabrieltseng) [inline tests/unit/eval/datasets/test_normalize.py:76]: same as above - this seems like it may warrant a warning ?

File: helios/evals/datasets/normalize.py (hunk 21 above)
Comment (Hgherzog) [reply helios/evals/datasets/normalize.py:56]: Fair, going to leave it in for now until we make a final decision
